### PR TITLE
Arm

### PR DESCRIPTION
**Pull request by save-buffer**

### CODE REVIEW

File: Arm/Arm.cs [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+            while (Angle < 0)
+                Angle += (float)(2 * Math.PI);
+            return Angle % (float)(2 * Math.PI);
+        }
+
+        public void MoveTo(float X, float Y, float Z)
+        {
+            const float LearningRate = 0.0001f;
+            const int Iterations = 40;
+            for (int j = 0; j < Params.Length; j++)
```
Comment (pensono) [inline Arm/Arm.cs]: I think this loop and the one that follows are swapped. We should be adjusting each parameter on every iteration.

File: Arm/Arm.cs [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+        ArmPart[] Params;
+        public float[] CurrentPitches { get; private set; }
+        public float[] CurrentRolls { get; private set; }
+        public float[] CurrentYaws { get; private set; }
+
+        private float[] TempPitches;
+        private float[] TempRolls;
+        private float[] TempYaws;
+
+        private float[] PitchSums;
```
Comment (pensono) [inline Arm/Arm.cs]: c# has a convention where private fields are lowercase and begin with an underscore to reduce confusion with property names.

File: Arm/Arm.cs [hunk 3]
```diff
@@ -0,0 +1,290 @@
+﻿using System;
+using System.Diagnostics;
+using System.Windows.Forms;
+using System.Drawing;
+
+namespace UseArm
+{
+    public struct ArmPart
```
[the remark this answers is not in the document]
Comment (save-buffer) [reply Arm/Arm.cs]: I chose struct to avoid double pointer dereference (since array dereference and then heap dereference)

File: Arm/Arm.cs [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+            this.MinPitch = MinPitch;
+            this.MaxPitch = MaxPitch;
+            this.MinYaw = MinYaw;
+            this.MaxYaw = MaxYaw;
+            Length = L;
+            Fixed = false;
+            FixedAngle = 0.0f;
+        }
+
+        public ArmPart(float FixedAngle, float Length)
```
[the remark this answers is not in the document]
Comment (save-buffer) [reply Arm/Arm.cs]: Pitch is fixed, but I agree. I will change it

File: Arm/Arm.cs [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+        public static implicit operator ArmPart((float, float, float, float, float, float, float) t)
+        {
+            return new ArmPart((float)t.Item1, (float)t.Item2, (float)t.Item3, (float)t.Item4, (float)t.Item5,
+                  (float)t.Item6, (float)t.Item7);
+        }
+    }
+
+    public class Arm
+    {
+        ArmPart[] Params;
```
[the remark this answers is not in the document]
Comment (save-buffer) [reply Arm/Arm.cs]: Default access modifier for members is private. I wasn’t explicit here, but I should be consistent with the others

File: Arm/Arm.cs [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+        public float[] CurrentPitches { get; private set; }
+        public float[] CurrentRolls { get; private set; }
+        public float[] CurrentYaws { get; private set; }
+
+        private float[] TempPitches;
```
[the remark this answers is not in the document]
Comment (save-buffer) [reply Arm/Arm.cs]: We need this because we need simultaneous updating of angles. Allocating an array inside the function call is still allocate on the heap. This way we only have to allocate once in constructor.

File: Arm/Arm.cs (hunk 1 above)
Comment (save-buffer) [reply Arm/Arm.cs]: I’ve noticed that this way performs much better because it updates the most impactful angles first. We can test it further, but from my preliminary testing it looked like this is better.

File: Arm/Arm.cs [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+
+        public Arm(params ArmPart[] Parameters)
+        {
+            this.Params = Parameters;
+            this.CurrentRolls = new float[Parameters.Length];
+            this.CurrentPitches = new float[Parameters.Length];
+            this.CurrentYaws = new float[Parameters.Length];
+
+            Random r = new Random();
+            for (int i = 0; i < Params.Length; i++)
```
[the remark this answers is not in the document]
Comment (save-buffer) [reply Arm/Arm.cs]: Randomization is needed because of everything is 0, gradients will be 0 and nothing will be updated. This is a common technique in machine learning, particularly neural networks.

File: Arm/Arm.cs [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+
+                        TempRolls[j] = CurrentRolls[j] - GradRoll;
+                        TempPitches[j] = CurrentPitches[j] - GradPitch;
+                        TempYaws[j] = CurrentYaws[j] - GradYaw;
+
+                        /*
+                        Console.WriteLine("GradRoll:  " + GradRoll);
+                        Console.WriteLine("GradPitch: " + GradPitch);
+                        Console.WriteLine("GradYaw:   " + GradYaw);
+                        */
```
Comment (BrianLunder) [inline Arm/Arm.cs]: Maybe delete this?

File: Arm/Arm.cs (hunk 2 above)
Comment (CaiB) [reply Arm/Arm.cs]: According to [Capitalization Conventions](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/capitalization-conventions):
> Do not use underscores to differentiate words, or for that matter, anywhere in identifiers.

And from [Names of Type Members](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/names-of-type-members):
> DO NOT use a prefix for field names. 

As far as I understand, everything should be PascalCase except for parameter names in methods, which should be camelCase.

File: Arm/Arm.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (save-buffer) [reply Arm/Arm.cs]: If we want to allocate the temporary arrays on the stack, we could allow unsafe code and use the stackalloc keyword. Probably not what we want to do.

File: Arm/Arm.cs [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+         * The front of the rover is along the X Axis
+         * The Y Axis is vertical, parallel to gravity
+         * The Z Axis is the cross product of the X and Y Axis
+         * 
+         * Here is how I'm defining the angles to be:
+         * Roll - Rotation about the X Axis
+         * Pitch - Rotation about the Z Axis
+         * Yaw - Rotation about the Y Axis
+         * This is contrary to norm. Usually Yaw is about the Z Axis
+         * and the Z Axis is parallel to gravity. Not here!
```
[the remark this answers is not in the document]
Comment (save-buffer) [reply Arm/Arm.cs]: Breaking from convention makes it easier to transition from 2D to 3D. If we want a 2D representation of the arm, we treat X as horizontal position and Y as vertical position. It's very natural, then, to add an extra axis (the Z axis) without changing anything else.

File: Arm/Arm.cs (hunk 1 above)
Comment (save-buffer) [reply Arm/Arm.cs]: I remembered the main rationale for me doing this while writing the code: it lets us clamp the angle after the gradient descent, meaning that the angle can move through illegal territory during gradient descent and only get clamped afterwards.
